smac/intensification/successive_halving.py

```diff
@@ -258,93 +259,248 @@ def intensify(self, challengers: typing.List[Configuration],
+
+    def next_challenger(self, challengers: typing.Optional[typing.List[Configuration]],
+                        chooser: typing.Optional['smac.optimizer.smbo.SMBO'],
+                        run_history: RunHistory,
+                        repeat_configs: bool = True) -> Configuration:
+        """
+        Selects which challenger to use based on the iteration stage and set the iteration parameters.
+        First iteration will choose configurations from the ``chooser`` or input challengers,
+        while the later iterations sample top configurations from the previously selected challengers in that iteration
 
+        Parameters
+        ----------
+        challengers : typing.List[Configuration]
+            promising configurations
+        chooser : 'smac.optimizer.smbo.SMBO'
+            optimizer that generates next configurations to use for racing
+        run_history : RunHistory
+            stores all runs we ran so far
+        repeat_configs : bool
+            if False, an evaluated configuration will not be generated again
+        """
+        # if this is the first run, then initialize tracking variables
+        if not hasattr(self, 'stage'):
+            self._update_sh_stage(run_history=run_history)
+        
+        curr_budget = int(self.all_budgets[self.stage])
+        prev_budget = int(self.all_budgets[self.stage - 1]) if self.stage > 0 else 0
+
+        # if all instances have been executed, then reset and move on to next config
+        n_insts = (curr_budget - prev_budget) if self.instance_as_budget else len(self.inst_seed_pairs)
+        n_insts_remaining = n_insts - self.curr_inst_idx
+
+        # if there are instances pending, finish running configuration
+        if self.running_challenger and n_insts_remaining > 0:
+            return self.running_challenger
+
+        # select next configuration
+        if self.stage == 0:
+            # first stage, so sample from configurations/chooser provided
+            challenger = self._next_challenger(challengers=challengers,
+                                               chooser=chooser,
+                                               run_history=run_history,
+                                               repeat_configs=repeat_configs)
+
+        else:
+            # sample top configs from previously sampled configurations
+            challenger = self.configs_to_run.pop()
+
+        if challenger:
+            # reset instance index for the new challenger
+            self.curr_inst_idx = 0
+            self._chall_indx += 1
+            self.running_challenger = challenger
+
+        return challenger
+
+    def _update_sh_stage(self, run_history: RunHistory = None) -> None:
+        """
+        Update tracking information for a new stage/iteration and update statistics
+
+        Parameters
+        ----------
+         run_history : RunHistory
+            stores all runs we ran so far
+        """
+
+        if not hasattr(self, 'stage'):
```

Comment:
Could you please do these in the init method? This would make it easier to understand the method above (if not reading this one first).